libs/ledger/ledger-common-data/src/main/kotlin/net/corda/ledger/common/data/transaction/Extensions.TransactionMetadata.kt

```diff
@@ -16,26 +16,56 @@ import net.corda.v5.crypto.merkle.HashDigestConstants
 import net.corda.v5.ledger.common.transaction.TransactionMetadata
 import java.util.*
 
-object TransactionMetadataUtils {
+object JsonMagic {
+    private val header = "corda".toByteArray() + byteArrayOf(8)
+    private const val OPENING_BRACKET = '{'.code.toByte()
+
+    /**
+     * returns a pair of schemaVersion to decoded json
+     * */
+    fun consume(byteData: ByteArray): Pair<Int?, String> {
+        val hasHeader = byteData.size > header.size && byteData.sliceArray(header.indices).contentEquals(header)
+        val hasNoHeader = byteData.isNotEmpty() && byteData[0] == OPENING_BRACKET
+
+        return if (hasHeader) {
+            byteData[header.size].toInt() to byteData.sliceArray(header.size + 1 until byteData.size).decodeToString()
+        } else if (hasNoHeader) {
+            // no header means schema version 1 which is compatible with platform version <= 5.2.1
+            1 to byteData.decodeToString()
+        } else {
+            null to ""
+        }
+    }
+}
 
+object TransactionMetadataUtils {
     fun parseMetadata(
         metadataBytes: ByteArray,
         jsonValidator: JsonValidator,
         jsonMarshallingService: JsonMarshallingService
     ): TransactionMetadataImpl {
-        val json = metadataBytes.decodeToString()
-        jsonValidator.validate(json, getMetadataSchema(jsonValidator))
-        val metadata = jsonMarshallingService.parse(json, TransactionMetadataImpl::class.java)
+        // extracting metadata schema version from a header and json.
+        val (version, json) = JsonMagic.consume(metadataBytes)
 
+        requireNotNull(version) {
+            "Metadata schema version is null."
```

Comment:
That is a "code" error, not what the user needs to know. If the version is null, the JSON blob was invalid, so rather do something like:
```suggestion
            "Transaction metadata JSON blob is invalid."
```